src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -227,34 +253,174 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
+                }
+
+                std::vector<std::pair<COORD, COORD>> matches;
+                if (!_searchState.value().Text.empty())
+                {
+                    const Search::Sensitivity sensitivity = _searchState.value().CaseSensitive ?
+                                                                Search::Sensitivity::CaseSensitive :
+                                                                Search::Sensitivity::CaseInsensitive;
+
+                    Search search(*GetUiaData(), _searchState.value().Text.c_str(), Search::Direction::Forward, sensitivity);
+                    while (co_await _SearchOne(search))
+                    {
+                        // if search box was collapsed or the new one search was triggered - let's cancel this one
+                        if (!_searchState.has_value() || _searchState.value().SearchId != originalSearchId)
+                        {
+                            co_return;
+                        }
+
+                        matches.push_back(search.GetFoundLocation());
+                    }
+
+                    // if search box was collapsed or the new one search was triggered - let's cancel this one
+                    if (!_searchState.has_value() || _searchState.value().SearchId != originalSearchId)
+                    {
+                        co_return;
+                    }
+                }
+                _searchState.value().Matches.emplace(matches);
+            }
+        }
+
+        if (auto control{ weakThis.get() })
         {
-            _terminal->SetBlockSelection(false);
-            search.Select();
+            _SelectSearchResult(goForward);
+        }
+    }
+
+    // Method Description:
+    // - Selects one of precomputed search results in the terminal (if exist).
+    // - Updates the search box control accordingly.
+    // - The selection might be preceded by going to next / previous result
+    // - goForward: if true, select next result; if false, select previous result;
+    // if not set, remain at the current result.
+    // Return Value:
+    // - <none>
+    void TermControl::_SelectSearchResult(std::optional<bool> goForward)
+    {
+        if (_searchState.has_value() && _searchState.value().Matches.has_value())
+        {
+            if (goForward.has_value())
+            {
+                _searchState.value().UpdateIndex(goForward.value());
+
+                const auto currentMatch = _searchState.value().GetCurrentMatch();
+                if (currentMatch.has_value())
+                {
+                    auto lock = _terminal->LockForWriting();
+                    _terminal->SetBlockSelection(false);
+                    _terminal->SelectNewRegion(currentMatch.value().first, currentMatch.value().second);
+                    _renderer->TriggerSelection();
+                }
+            }
+
+            if (_searchBox)
+            {
+                _searchBox->SetStatus(_searchState.value().Status());

```

Comment:
do we need a `co_await winrt::resume_foreground(...)` here? Presumably `_SearchAsync` could collect matches on the BG thread, then call this function, which will need to update the UI...

Reply:
(just saw now that `_SearchOne` returns to the FG thread before returning the result)